File: 20131209/AM/131209_tests/subway/check/subway3.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <string>
#include <vector>

using namespace std;

string canonical(const string& s) {
  vector<vector<string> > stack;
  stack.push_back(vector<string>());

  for (size_t i = 0; i < s.size(); ++i) {
    if (s[i] == '0') {
      stack.push_back(vector<string>());
    } else {
      vector<string>& children = stack[stack.size()-1];
      string sortedChildren;
      sort(children.begin(), children.end());
      for (size_t j = 0; j < children.size(); ++j) {
        sortedChildren += "0";
        sortedChildren += children[j];
        sortedChildren += "1";
      }

      stack.pop_back();
      stack[stack.size()-1].push_back(sortedChildren);
    }
  }

  string ans;
  sort(stack[0].begin(), stack[0].end());

  for (size_t i = 0; i < stack[0].size(); ++i) {
    ans += "0";
    ans += stack[0][i];
    ans += "1";
  }

  return ans;
}

int main(int argc, char** argv) {
  int n;
  cin >> n;

  set<string> unique;

  for (int i = 0; i < n; ++i) {
    string s;
    cin >> s;

    unique.insert(canonical(s));
  }

  cout << unique.size() << endl;
}

File: 20131209/AM/131209_tests/subway/check/subway3.java
import java.io.*;
import java.util.*;

public class subway {
  public static String canonical(String s) {
    Vector<Vector<String>> stack = new Vector<Vector<String>>();
    stack.add(new Vector<String>());

    for (int i = 0; i < s.length(); ++i) {
      if (s.charAt(i) == '0') {
        stack.add(new Vector<String>());
      } else {
        Vector<String> children = stack.get(stack.size()-1);
        String sortedChildren = "";
        Collections.sort(children);
        for (int j = 0; j < children.size(); ++j) {
          sortedChildren += "0";
          sortedChildren += children.get(j);
          sortedChildren += "1";
        }

        stack.remove(stack.size()-1);
        stack.get(stack.size()-1).add(sortedChildren);
      }
    }

    String ans = "";
    Collections.sort(stack.get(0));

    for (int i = 0; i < stack.get(0).size(); ++i) {
      ans += "0";
      ans += stack.get(0).get(i);
      ans += "1";
    }

    return ans;
  }

  public static void main(String[] args) {
    Scanner fin = new Scanner(System.in);
    int n = fin.nextInt();

    Set<String> unique = new TreeSet<String>();

    for (int i = 0; i < n; ++i) {
      String s = fin.next();

      unique.add(canonical(s));
    }

    System.out.println(unique.size());
  }
}
